### PR TITLE
parser: use new param metadata API for special data

### PR DESCRIPTION
**Pull request by rwcarlsen**
The parser previously injected extra metadata about parsed parameters
into InputParameters objects as specially named parameters.  A recent
refactor of the InputParameters class made it desirable to change this
to use an explicit param metadata API approach.  This is the final
enabling change in preparation for addressing #10357.

### CODE REVIEW

**Comment by rwcarlsen** [comment]
A bison submodule update for sabertooth should fix that failure.

**Comment by permcody** [comment]
Sounds good, go ahead and push one, I can merge.

**Comment by rwcarlsen** [comment]
All green.

File: framework/include/utils/InputParameters.h
Comment (permcody) [inline framework/include/utils/InputParameters.h]: Per our discussion yesterday. Let's remove these unchecked `at()` methods from the code. The problem is that if they are called and we don't catch the exception a rank may mysteriously die leaving the user with no error message and instead they receive a cryptic error about MPI running into some problem. It's super nasty to track down. 

@friedmud suggested deriving from standard map and doing simple error checking on the `at()` method but that would be pretty invasive in that we'd have to force people to stop using std::map everywhere. It's not a terrible solution but it's just one more things that users have to look at and wonder why or alternatively run into error messages with CIVET if we enforce that check. I'll think about this some more.

In the meantime, there's nothing stopping you from creating a simple macro if you'd like, or just avoid using that method. I'd entertain either.

**Comment by rwcarlsen** [comment]
I removed the unprotected ``map.at`` calls.  Should be ready for a merge now.

File: framework/include/utils/InputParameters.h
Comment (permcody) [inline framework/include/utils/InputParameters.h]: Note: We normally don't use C-style casts anywhere in MOOSE. Since this line of code is temporary, I'll let it slide. In the future, use the (longer) C++-style casts.

File: framework/include/utils/InputParameters.h
Comment (permcody) [inline framework/include/utils/InputParameters.h]: This isn't the correct format for comments in MOOSE source files. We use the block format for methods and three-slashes for member variables. Also, so that Doxygen knows that two methods share the same comment you'll want to use the special "///@{ ///@}" syntax around these paired methods. Let's get in the habit of getting this correct. It makes the maintenance easier if we all follow the same conventions. If you make only comment changes, we won't need to wait for the testing cycle to complete before this can be merged.

**Review by permcody — CHANGES_REQUESTED** [review]

**Comment by rwcarlsen** [comment]
Comments fixed.

**Review by permcody — APPROVED** [review]

**Pull request merged**
